File: src/omnisight/presentation/static/js/core/bus.js
// 轻量事件总线。只服务"跨模块的一次性通知"（按键动画、数据失效），
// 状态一律走 store——两条通道混用会让"现在到底谁是真相"变得不可回答。
/** @type {Map<string, Set<(payload: any, topic: string) => void>>} */
const topics = new Map();

/**
 * @param {string} topic
 * @param {(payload: any, topic: string) => void} handler
 * @returns {() => void} 注销函数
 */
export function on(topic, handler) {
  let handlers = topics.get(topic);
  if (!handlers) {
    handlers = new Set();
    topics.set(topic, handlers);
  }
  handlers.add(handler);
  return () => off(topic, handler);
}

/**
 * @param {string} topic
 * @param {(payload: any, topic: string) => void} handler
 */
export function off(topic, handler) {
  topics.get(topic)?.delete(handler);
}

/**
 * @param {string} topic
 * @param {unknown} [payload]
 */
export function emit(topic, payload) {
  const handlers = topics.get(topic);
  if (!handlers) return;
  // 复制一份再遍历：处理器里注销自己是常见写法。
  for (const handler of [...handlers]) {
    try {
      handler(payload, topic);
    } catch (error) {
      console.error(`总线处理器失败：${topic}`, error);
    }
  }
}

export function clear() {
  topics.clear();
}

File: src/omnisight/core/bus.py
"""进程内事件总线：采集侧发布、SSE 与托盘订阅。

刻意做得很小：没有通配符、没有优先级、没有异步。它唯一的职责是让"按键落地"
与"页面收到通知"解耦，而 SSE 的实际扇出在 05 文档 §7 里另有节流逻辑。

订阅者回调中的异常**不允许**冒泡到发布者——发布者往往是采集热路径，一个坏掉的
订阅者不该让采集停摆。
"""

from __future__ import annotations

import logging
import threading
from collections.abc import Callable
from typing import Any

logger = logging.getLogger(__name__)

Listener = Callable[[str, Any], None]


class EventBus:
    __slots__ = ("_listeners", "_lock")

    def __init__(self) -> None:
        self._listeners: dict[str, list[Listener]] = {}
        self._lock = threading.Lock()

    def subscribe(self, topic: str, listener: Listener) -> Callable[[], None]:
        """订阅并返回取消订阅的可调用对象。"""
        with self._lock:
            self._listeners.setdefault(topic, []).append(listener)

        def unsubscribe() -> None:
            with self._lock:
                handlers = self._listeners.get(topic)
                if handlers and listener in handlers:
                    handlers.remove(listener)

        return unsubscribe

    def publish(self, topic: str, payload: Any = None) -> None:
        with self._lock:
            handlers = list(self._listeners.get(topic, ()))
        for handler in handlers:
            try:
                handler(topic, payload)
            except Exception:
                logger.exception("事件订阅者处理 %s 时抛出异常，已忽略", topic)

    def topics(self) -> tuple[str, ...]:
        with self._lock:
            return tuple(self._listeners)
